_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+                s"otpauth://${generator.toLowerCase()}/" +
+                URLEncoder.encode(issuer + ":" + account, "UTF-8").replace("+","%20") +
+                "?secret=" + URLEncoder.encode(secretKey, "UTF-8").replace("+","%20") +
+                "&issuer=" + URLEncoder.encode(issuer, "UTF-8").replace("+","%20") +
+                "&conter=0"
+                }
+            case "Totp" =>
+                Try{
+                    s"otpauth://${generator.toLowerCase()}/" +
+                    URLEncoder.encode(issuer + ":" + account, "UTF-8").replace("+","%20") +
+                    "?secret=" + URLEncoder.encode(secretKey, "UTF-8").replace("+","%20") +
+                    "&issuer=" + URLEncoder.encode(issuer, "UTF-8").replace("+","%20")
+                }
+end BarCodeService
+```
+
+The `getGoogleAuthenticatorBarCode()` method takes a `secretKey` which we create with `otp-java`, an `account` and `issuer` which are the email address and name of the business or person issuing the barcode, and a `generator` which is a string of either `Hotp` or `Totp` (used to choose the type of `otp`), this gives us a `Try[String]`. This function creates a `URI` in the format `otpauth://TYPE/LABEL?PARAMETERS` that we'll use to create the QR Bar Code image depending on the type of `generator` value we pass. The `Hotp` implementation contains an extra parameter, `counter=0`.
+
+```scala
+import com.google.zxing.common.BitMatrix
+import com.google.zxing.MultiFormatWriter
+import com.google.zxing.BarcodeFormat
+import java.io.FileOutputStream
+import com.google.zxing.client.j2se.MatrixToImageWriter
+
+trait BarCodeService:
+    ...
+    def createQRCode(barCodeData: String, filePath: String, height: Int, width: Int) =
+        val matrix: BitMatrix = new MultiFormatWriter().encode(barCodeData, BarcodeFormat.QR_CODE, width, height)
+        val outVal: FileOutputStream = new FileOutputStream(filePath)
+        MatrixToImageWriter.writeToStream(matrix,"png",outVal)
+end BarCodeService
+```
+
+The `createQRCode()` function creates the BarCode image. It takes as parameters, `barCodeData` which is our `URI`, a `filePath`, and `height` and `width` of our image. The `MultiFormatWriter` class has an encode method that takes our arguments and creates a 2D matrix of bits.
+The `writeToStream` method on the `MatrixToImageWriter` object takes the `matrix` (of type `BitMatrix`), `"png"` as the image format and a `FileOutputStream`, writes the image to a png file and saves it to the file path.
+
+Here's the full code
+
+```scala
+package com.rockthejvm
+
+import scala.util.*
+import java.net.URLEncoder
+import com.google.zxing.common.BitMatrix
+import com.google.zxing.MultiFormatWriter
+import com.google.zxing.BarcodeFormat
+import java.io.FileOutputStream
+import com.google.zxing.client.j2se.MatrixToImageWriter
+
+trait BarCodeService:
+    def getGoogleAuthenticatorBarCode(secretKey: String, account: String, issuer: String, generator: String): Try[String] =
```

Comment:
why are we returning a Try here? seems unnecessary

Reply:
Switched to Either